### PR TITLE
Rtp h265

### PR DESCRIPTION
**Pull request by rakeshnitb**

### CODE REVIEW

**Comment by tonihei** [comment]
@claincly Could you take a look at this one as well? Thanks!

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+  /** Offset of payload data within a FU type A payload. */
+  private static final int FU_PAYLOAD_OFFSET = 3;
+
+  /** Single Time Aggregation Packet type A. */
+  private static final int RTP_PACKET_TYPE_STAP_A = 48; // RFC7798 Section 4.4.2
+  /** Fragmentation Unit type A. */
+  private static final int RTP_PACKET_TYPE_FU_A = 49;
+
+  /** IDR NAL unit type. */
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:53]: I don't see IDR_W_LP in H265 spec or RFC7798, maybe I missed something? Or could it be IDR_W_RADL?

It'd be useful if a link is attached.

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  private final RtpPayloadFormat payloadFormat;
+
+  private @MonotonicNonNull TrackOutput trackOutput;
+  @C.BufferFlags private int bufferFlags;
+
+  private long firstReceivedTimestamp;
+  private int previousSequenceNumber;
+  /** The combined size of a sample that is fragmented into multiple RTP packets. */
+  private int fragmentedSampleSizeBytes;
+
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:73]: Super-nit: remove the blank lines L68 and 73

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    this.payloadFormat = payloadFormat;
+    fuScratchBuffer = new ParsableByteArray();
+    firstReceivedTimestamp = C.TIME_UNSET;
+    previousSequenceNumber = C.INDEX_UNSET;
+  }
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, int trackId) {
+    trackOutput = extractorOutput.track(trackId, C.TRACK_TYPE_VIDEO);
+
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:87]: Super-nit: remove the blank line

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    castNonNull(trackOutput).format(payloadFormat.format);
+  }
+
+  @Override
+  public void onReceivingFirstPacket(long timestamp, int sequenceNumber) {}
+
+  @Override
+  public void consume(ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker)
+      throws ParserException {
+
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:97]: Super-nit: remove the blank line

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    trackOutput.sampleData(data, numBytesInData);
+    fragmentedSampleSizeBytes += numBytesInData;
+
+    int nalHeaderType = (data.getData()[0] >> 1) & 0x3F;
+    bufferFlags = getBufferFlagsFromNalType(nalHeaderType);
+  }
+
+
+  /**
+   * Processes STAP Type A packet (RFC7798 Section 4.4.2).
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:178]: Processes an AP packet (RFC7798 Section 4.4.2).

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    //    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //    | NALU 2 HDR    |                                               |
+    //    +-+-+-+-+-+-+-+-                    NALU 2 Data                 |
+    //    |                                                               |
+    //    |                               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //    |                               :...OPTIONAL RTP padding        |
+    //    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+    throw ParserException.createForMalformedManifest(
+        "need to implement processSingleTimeAggregationPacket",
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:209]: Do you have an ETA for this?

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        /* cause= */ null);
+
+  }
+
+  /**
+   * Processes Fragmentation Unit Type A packet (RFC7798 Section 4.4.3).
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:215]: Processes Fragmentation Unit packet (RFC7798 Section 4.4.3). (drop type a)

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    //   |                                                               |
+    //   |                               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //   |                               :...OPTIONAL RTP padding        |
+    //   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //   FU header
+    //   +---------------+
+    //   |0|1|2|3|4|5|6|7|
+    //   +-+-+-+-+-+-+-+-+
+    //   |S|E|  FuType   |
+    //   +---------------+
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:243]: Maybe include the payloadHdr diagram? 

```
            +---------------+---------------+
            |0|1|2|3|4|5|6|7|0|1|2|3|4|5|6|7|
            +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
            |F|   Type    |  LayerId  | TID |
            +-------------+-----------------+

```

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+    int tid = (data.getData()[1] & 0x7); // last 3 bits in byte 1 of payload header section 1.1.4
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:245]: RFC7798 Section 1.1.4, ditto on line 250 and 251, please note the case.

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    int fuHeader = data.getData()[2];
+    int nalUnitType = fuHeader & 0x3F;
+    byte nalHeader[] = new byte[2];
+
+    nalHeader[0] = (byte) (nalUnitType << 1); // Section: 1.1.4
+    // According to section 1.1.4 in rfc7798, layerId is required to be zero so keeping its value
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:251]: Nit: Maybe "layerId must be zero according to RFC 7798 Section 1.1.4, so copying the tid only"

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+    // zero and copying only tid.
+    nalHeader[1] = (byte) tid;
+    boolean isFirstFuPacket = (fuHeader & 0x80) > 0;
+    boolean isLastFuPacket = (fuHeader & 0x40) > 0;
+
+    if (isFirstFuPacket) {
+      // Prepends starter code.
+      fragmentedSampleSizeBytes += writeStartCode();
+
+      // The bytes needed is 2 (NALU header) + payload size. The original data array has size 3
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:261]: The bytes needed is 2 (NALU header**s**)

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+      // (2 payload + 1 FU header) + payload size. Thus setting the correct header and set position
+      // to 1.
+      // Overwrite byte 1 of payload header with byte 0 of HEVC nal header
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:264]: NAL header

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+      data.getData()[1] = (byte) nalHeader[0];
+      // Overwrite byte FU Header with byte 1 of HEVC nal header
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:266]: Maybe missed something between "Overwrite" and "byte"?

Also, HEVE NAL header.

**Review by claincly — CHANGES_REQUESTED** [review]

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java (hunk 6 above)
Comment (claincly) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java]: I'm not sure "not finding a server to test" is a valid reason that you "can't implement it". Please provide an implementation following the RFC, and if possible, do some basic tests like feeding some human generated packets to it.

Comment (rakeshnitb) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java]: Added a TODO for it, we will visit again to complete it.

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+  @C.BufferFlags private int bufferFlags;
+  private long firstReceivedTimestamp;
+  private int previousSequenceNumber;
+  /** The combined size of a sample that is fragmented into multiple RTP packets. */
+  private int fragmentedSampleSizeBytes;
+  private long startTimeOffsetUs;
+
+  /** Creates an instance. */
+  public RtpH265Reader(RtpPayloadFormat payloadFormat) {
+    this.payloadFormat = payloadFormat;
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:87]: super-nit: please move this below line 88 to match the order they are declared.

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+
+  /** IDR NAL unit types. */
+  private static final int NAL_IDR_W_RADL = 19;
+  private static final int NAL_IDR_N_LP = 20;
+
+  /** Scratch for Fragmentation Unit RTP packets. */
+  private final ParsableByteArray fuScratchBuffer;
+
+  private final ParsableByteArray nalStartCodeArray =
+      new ParsableByteArray(NalUnitUtil.NAL_START_CODE);
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:73]: super-nit: initialize this in the constructor
super-nit: remove the spaces between private final fields

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+    //   |0|1|2|3|4|5|6|7|0|1|2|3|4|5|6|7|
+    //   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //   |F|   Type    |  LayerId  | TID |
+    //   +-------------+-----------------+
+
+    int tid =
+        (data.getData()[1] & 0x7); // last 3 bits in byte 1 of payload header, RFC7798 Section 1.1.4
+    int fuHeader = data.getData()[2];
+    int nalUnitType = fuHeader & 0x3F;
+    byte nalHeader[] = new byte[2];
```
Comment (claincly) [inline libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:238]: I feel we can skip creating this small array, by merging L240 and 256, L242 and 257. 

given this method is called very frequently.

**Review by claincly — CHANGES_REQUESTED** [review]
mostly nits

**Review by claincly — APPROVED** [review]

File: libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+      // Prepends starter code.
+      fragmentedSampleSizeBytes += writeStartCode();
+
+      // Overwrite a few bytes in Rtp buffer to get HEVC NAL Unit
+      // Rtp Byte 0  -> Ignore
+      // Rtp Byte 1  -> nal_unit_type, RFC7798 Section 1.1.4
+      // Rtp Byte 2  -> layerId required to be zero so copying only tid, RFC7798 Section 1.1.4
+      // Rtp Payload -> HEVC NAL bytes, so leave them unchanged
+      // Set data position from byte 1 as byte 0 was ignored
+      data.getData()[1] = (byte) (nalUnitType << 1);
```
[the remark this answers is not in the document]
Comment (rakeshnitb) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:250]: @claincly Here i'm parsing FU header to get nal_unit_header and setting it in HEVC NAL Header format. nal_unit_type is of 6 bits from bit 2 to 7 in FU header, so i'm doing (fuHeader & 0x3F) to get nal_unit_type. In HEVC NAL unit Header nal_unit_header is from bit 1 to 7 so i'm doing (nalUnitType << 1).

[the remark this answers is not in the document]
Comment (claincly) [reply libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpH265Reader.java:250]: Thanks, I missed a bit here. All good then.

**Comment by claincly** [comment]
Hi Rakesh, this PR is submitted now and should be reflected here once we do a push. 

The main difference is we removed the empty method `processAggregationPacket` and moved the TODO/throw to its call site.

Also, please write some tests for the fragmentation unit processing. It can be as simple as `RtpAc3ReaderTest.java`.

**Pull request merged**